service/history/queue/cross_cluster_queue_processor_base.go

```diff
@@ -0,0 +1,688 @@
+				return task.NewCrossClusterSignalWorkflowTask(
+					shard,
+					taskInfo,
+					task.InitializeLoggerForTask(shard.GetShardID(), taskInfo, logger),
+					shard.GetTimeSource(),
+					shard.GetConfig().CrossClusterTaskMaxRetryCount,
+				)
+			case persistence.CrossClusterTaskTypeCancelExecution:
+				return task.NewCrossClusterCancelWorkflowTask(
+					shard,
+					taskInfo,
+					task.InitializeLoggerForTask(shard.GetShardID(), taskInfo, logger),
+					shard.GetTimeSource(),
+					shard.GetConfig().CrossClusterTaskMaxRetryCount,
+				)
+			default:
+				panic(fmt.Sprintf("received unsupported cross cluster task type: %v", taskInfo.GetTaskType()))
+			}
+		},
+
+		backoffTimer:  make(map[int]*time.Timer),
+		shouldProcess: make(map[int]bool),
+
+		notifyCh:  make(chan struct{}, 1),
+		processCh: make(chan struct{}, 1),
+	}
+
+	return crossClusterQueueProcessorBase
+}
+
+func (c *crossClusterQueueProcessorBase) Start() {
+	if !atomic.CompareAndSwapInt32(&c.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+
+	c.logger.Info("Cross cluster queue processor state changed", tag.LifeCycleStarting)
+	defer c.logger.Info("Cross cluster queue processor state changed", tag.LifeCycleStarted)
+
+	c.redispatcher.Start()
+
+	// trigger an initial load of tasks
+	c.notifyAllQueueCollections()
+
+	c.shutdownWG.Add(1)
+	go c.processLoop()
+}
+
+func (c *crossClusterQueueProcessorBase) Stop() {
+	if !atomic.CompareAndSwapInt32(&c.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+
+	c.logger.Info("Cross cluster queue processor state changed", tag.LifeCycleStopping)
+	defer c.logger.Info("Cross cluster queue processor state changed", tag.LifeCycleStopped)
+
+	close(c.shutdownCh)
+	c.processingLock.Lock()
+	for _, timer := range c.backoffTimer {
+		timer.Stop()
+	}
+	for level := range c.shouldProcess {
+		c.shouldProcess[level] = false
+	}
+	c.processingLock.Unlock()
+
+	if success := common.AwaitWaitGroup(&c.shutdownWG, time.Minute); !success {
+		c.logger.Warn("", tag.LifeCycleStopTimedout)
+	}
+
+	c.redispatcher.Stop()
+}
+
+func (c *crossClusterQueueProcessorBase) processLoop() {
+	defer c.shutdownWG.Done()
+
+	updateAckTimer := time.NewTimer(backoff.JitDuration(
+		c.options.UpdateAckInterval(),
+		c.options.UpdateAckIntervalJitterCoefficient(),
+	))
+	defer updateAckTimer.Stop()
+
+	maxPollTimer := time.NewTimer(backoff.JitDuration(
+		c.options.MaxPollInterval(),
+		c.options.MaxPollIntervalJitterCoefficient(),
+	))
+	defer maxPollTimer.Stop()
+
+	var ackLevel int64
+	lastPollTime := time.Now()
+processorPumpLoop:
+	for {
+		select {
+		case <-c.shutdownCh:
+			break processorPumpLoop
+		case <-c.notifyCh:
+			c.notifyAllQueueCollections()
+		case <-updateAckTimer.C:
+			processFinished, err := c.updateAckLevel()
+			if err == shard.ErrShardClosed || (err == nil && processFinished) {
+				go c.Stop()
+				break processorPumpLoop
+			}
+			ackLevel, err = c.completeAckTasks(ackLevel)
```

Comment:
- Looks like we can change `updateAckLevel()` to also return the new ackLevel and pass it into `completeAckTasks()` so that we don't have to compute it again? 
- The current ackLevel can maybe be a field of `crossClusterQueueProcessorBase`? Currently, it's kind of confusing to me that the ackLevel pass into the function is actually the old ack level that is not used for deleting tasks and a new one is calculated inside the func.

Reply:
This is going to be a change that affects all three queue types. I added a TODO and will do it in the following PR

Reply:
I included the change in this PR